matrix-org/synapse
# Support oEmbed for media previews.

Fixes #7643.

This fixes Twitter previews by querying for oEmbed information instead of the originally requested URL. It is mostly generic (so additional oEmbed endpoints can be added), but it isn't end-user configurable at the moment.

The gist of the changes to the preview algorithm is:

1. Check if the requested URL matches any of the oEmbed patterns, if so:
    1. Query the oEmbed endpoint for the JSON metadata.
    2. If the metadata is a photo or has a thumbnail, use that instead of the original URL and continue the normal preview code.
    3. If the metadata is HTML, use the HTML directly.

A big section of `_download_url` got indented, but was otherwise unchanged.

A couple of questions I have:
* Does it make sense to make this user configurable?
* Parsing of the globs to regular expressions at module load time seems kind of meh, but is easy. I wonder if we should have a separate handler or something for oEmbed?
* Are there other services we want to add? (Note that there's a JSON file which contains all the providers, we could fetch it programmatically or include it.)

Note that this does NOT attempt to support "discovery" via `<link>` tags. (Nor does it handle XML.)

## Testing steps

1. Enable URL previews in your homeserver (I think this involes `enable_media_repo`, `url_preview_enabled`, `url_preview_ip_range_blacklist`).
2. Join a room.
3. Enable URL previews for that room in the settings.
4. Send a link that should get previewed and make sure it still works.
5. Send a Twitter link to get previewed (e.g. https://twitter.com/matrixdotorg/status/1285937589791399936) and see a preview text.
